assets/src/vo/date-time/duration.js

```diff
@@ -0,0 +1,567 @@
+	}
+
+	/**
+	 * Create an instance of Duration from a simple object.
+	 *
+	 * @param {Object} values  Keys should be the units (eg 'years', 'days').
+	 * @param {string} locale
+	 * @return {Duration} An instance of Duration
+	 */
+	static fromObject( values, locale = DEFAULT_VALID_LOCALE ) {
+		return new Duration( values, locale );
+	}
+
+	/**
+	 * Create an instance of Duration from an ISO8601 string.
+	 *
+	 * @param {string} ISOString (eg. 'PT23H')
+	 * @param {string} locale
+	 * @return {Duration} An instance of Duration
+	 */
+	static fromISO( ISOString, locale = DEFAULT_VALID_LOCALE ) {
+		assertions.assertISO8601IsValid( ISOString, true );
+		return new Duration( ISOString, locale );
+	}
+
+	/**
+	 * Indicates whether the provided value is a valid instance of Duration.
+	 * @param {mixed|Duration}duration
+	 * @return {boolean}  True means it is a valid Duration object.
+	 */
+	static isValidDuration( duration ) {
+		return duration instanceof Duration && duration.isValid;
+	}
+
+	/**
+	 * Asserts whether the provided value is a valid Duration and throws an
+	 * exception if not.
+	 * @param {mixed|Duration} duration
+	 * @throws TypeError
+	 */
+	static assertIsValidDuration( duration ) {
+		if ( ! Duration.isValidDuration( duration ) ) {
+			throw new TypeError(
+				'This Duration object is not valid.'
+			);
+		}
+	}
+
+	/**
+	 * Indicates whether the provided value is an instance of Duration.
+	 *
+	 * @param {Duration|mixed} duration
+	 * @return {boolean}  True means the value is an instance of Duration.
+	 * Note: true may still mean that the Duration instance is not valid!
+	 */
+	static isDuration( duration ) {
+		return duration instanceof Duration;
+	}
+
+	/**
+	 * Asserts whether the provided value is an instance of Duration and if not
+	 * throws an exception.
+	 *
+	 * @param {Duration|mixed} duration
+	 * @throws TypeError
+	 */
+	static assertIsDuration( duration ) {
+		if ( ! Duration.isDuration( duration ) ) {
+			throw new TypeError(
+				'The provided value is not an instance of Duration.'
+			);
+		}
+	}
+
+	/**
+	 * This filters the incoming values and returns only key/value pairs that
+	 * are acceptable as duration units.
+	 *
+	 * If a invalid duration unit is discovered, a console.error is generated
+	 * (in non-production mode).
+	 *
+	 * @param {mixed} values
+	 * @return {Object} Filtered values.
+	 * @throws TypeError if incoming values argument is not an object.
+	 */
+	[ privateMethods.filterValues ]( values ) {
+		if ( typeof values !== 'object' ) {
+			throw new TypeError( 'Incoming values must be a simple object.' );
+		}
+		const valuesToSet = pick( values, unitNames );
+		if ( ! isShallowEqual( values, valuesToSet ) ) {
+			warning(
+				false,
+				'The following unexpected keys were in the configuration ' +
+				'object for constructing the Duration: ' +
+				keys( omit( values, unitNames ) ).join()
+			);
+			this[ privateProperties.isValid ] = false;
```

Comment:
So the only time we can know whether something is valid or not is _when_ our `Duration` is constructed with a simple object.  Anything else goes directly to `moment.duration()` and although `moment.duration` has a `isValid()` method, it returns true for anything passed into the constructor.  Try it!  Go to the momentjs.com website (which has moment pre-loaded) and try to get an invalid result:

```js
dur = moment.duration( 'invalid' );
dur.isValid() // true

dur = moment.duration( new Date() );
dur.isValid() //true

dur = moemnt.duration( 1, 'cheeseburgers' );
dur.isValid() //true
```

So, `isValid` is only usable for when a duration is constructed with a simple Object and thus we do our validation on it at that point.

Reply:
I think there's some sort of miscommunication happening here.  If I don't have our own `privateProperties.isValid` method then effectively our Duration object will never be invalid because as I was trying to demonstrate via the examples that constructing `moment.duration` from ANY incoming value will never cause `moment.duration.isValid()` to return false!

Since we're saving incoming simple object property values to its own isolated object (separate from moment), and we _can_ validate that object before passing it along to moment, I opted to validate what we could and use that to set whether the object is valid or not.  I think that's better than not doing anything.  Eventually if/when we switch to another internal duration library (such as Luxon's Duration) it's entirely possible at that point that the library would consider something like `{cheeseburgers: 10}` to be invalid and return so for its own `isValid()` method.  However `moment.duration` does not.

Reply:
your examples:

```js
dur = moment.duration( 'invalid' );
dur.isValid() // true
dur = moment.duration( new Date() );
dur.isValid() // true
dur = moemnt.duration( 1, 'cheeseburgers' );
dur.isValid() // true
```

were followed immediately by this comment:

> So, isValid is only usable for when a duration is constructed with a simple Object and thus we do our validation on it at that point.

but I missed that you switched from `isValid()` in the examples to immediately referring to `isValid` the property, so it appeared that you were implying that the following would result in invalid duration objects:

```js
dur = moment.duration( { 'invalid' } );
dur.isValid() // false

dur = moment.duration( { new Date() } );
dur.isValid() // false

dur = moemnt.duration( { cheeseburgers: 2 } );
dur.isValid() // false
```

ie: I thought you were saying:

> So, isValid() is only usable for when a duration is constructed with a simple Object and thus we do our validation on it at that point.

which to me sounded like you were saying that `isValid()` worked correctly if you supplied duration with simple objects (but not other parameter types as demonstrated by your examples), which made validating via an internal `isValid` property non-sensical since it worked exactly the same way. Make sense? So my bad for interpreting things wrong.

Anyways... a couple of points...

- a possible reason why `isValid()` may not work correctly with durations is because it is a method on the moment object but not a method on the duration object
- although `isValid()` returns true for the examples above, `duration.toISOString()` seems to reliably return `"P0D"` for any garbage that you throw at it, so we could maybe use that as a validation check at the end of our constructor, ie:

```js
// check that duration is not zero
if ( this[ privateProperties.duration ].toISOString() === 'P0D' ) {
    this[ privateProperties.isValid ] = false;
}
```

which would not only work for ALL types of construction ( number, string, simple object ) but could also remove the need for the `filterValues()` method, although I think that it would still be good for notifying users when there is a problem with their property keys. At the very least, it would not need to toggle the `isValid` property since that would happen anyways via the above check.